### PR TITLE
Feature/meta exponential

### PR DESCRIPTION
**Pull request by adolgert**
This is a set of tools with which to build modules that do discrete time steps. It's a piece of that puzzle, the most mathy part, written in a way that should help get the math right.

### CODE REVIEW

File: macro/vignettes/design_meta_timestep.Rmd [hunk 1]
```diff
@@ -0,0 +1,403 @@
...
+X ~ h * U - sin(r) * X
+)
+```
+I'll assume this is a differential form of an ODE that reflects a master equation.
+That is, I'll assume that everything that leaves one compartment enters another compartment
+and that there are no constant terms.
+
+**Are these limitations the best choices?**
+
+From this kind of specification, we can produce lots of different kinds of time steps.
```
Comment (slwu89) [inline macro/vignettes/design_meta_timestep.Rmd:36]: I'm not sure how this sentence answers/relates to the question raised in the bold text.

File: macro/vignettes/design_meta_timestep.Rmd [hunk 2]
```diff
@@ -0,0 +1,403 @@
...
+
+# Discrete Time Steps
+
+## The Matrix for the Master Equation
+Most of them are simple permutations of the specification. The simplest is a function
```
Comment (slwu89) [inline macro/vignettes/design_meta_timestep.Rmd]: I don't understand what `Most of them are simple permutations of the specification.` means.

File: macro/vignettes/design_meta_timestep.Rmd [hunk 3]
```diff
@@ -0,0 +1,403 @@
...
+```{r run_small}
+si_small(list(h = 0.4, r = 0.1, t = 0.05))
+```
+
+How do you use this?
+The small time step is a very common way to build steps. We think
+of rate times time as the marginal probability for a jump in a Markov
+chain. If parameters keep changing, then this function can build the matrix quickly.
+If you wanted to simulate a long interval that has constant
+parameters, you could split it into $2^(2n)$ steps and square this matrix
```
Comment (slwu89) [inline macro/vignettes/design_meta_timestep.Rmd]: formatting got weird in PDF, maybe try $2^{(2n)}$

File: macro/vignettes/design_meta_timestep.Rmd [hunk 4]
```diff
@@ -0,0 +1,403 @@
...
+```
+We can execute it by passing in the original parameters, plus a time.
+```{r run_small}
+si_small(list(h = 0.4, r = 0.1, t = 0.05))
+```
+
+How do you use this?
+The small time step is a very common way to build steps. We think
+of rate times time as the marginal probability for a jump in a Markov
+chain. If parameters keep changing, then this function can build the matrix quickly.
```
Comment (slwu89) [inline macro/vignettes/design_meta_timestep.Rmd]: I don't understand what it means that this function can build the matrix quickly when parameters keep changing.

File: macro/vignettes/design_meta_timestep.Rmd [hunk 5]
```diff
@@ -0,0 +1,403 @@
...
+
+## Exact Exponential
+
+It is possible to construct the rate matrix, $At$, and exponentiate it, in order
+to get an exact solution to the time step. The time step is a numerical calculation
+of the expontial in
+$$
+x_{t+1} = e^{At} x_t.
+$$
+It's as good as integrating, and R
```
Comment (slwu89) [inline macro/vignettes/design_meta_timestep.Rmd]: I might be even a little more pedantic here and say "It's as good as integrating the master equation (ODEs) component by component" or something like that.

File: macro/vignettes/design_meta_timestep.Rmd [hunk 6]
```diff
@@ -0,0 +1,403 @@
...
+
+## Single Jump
+
+The single jump time step is an exact solution to the master equation, assuming
+that an individual can make only one jump per time step. If there is a compartment,
+$S$, and that can go to compartments $E_0$, $E_1$, $E_2$ with hazard rates
+$\lambda_1$, $\lambda_2$, $\lambda_3$, then the probability
+of remaining in $S$ is
+$$
+  P[S_{t+1}|S_t] = e^{\sum_i\lambda_i}
```
Comment (slwu89) [inline macro/vignettes/design_meta_timestep.Rmd]: Should be `e^{-\sum_i\lambda_i}`, also maybe remark this is for a unit length time step.

File: macro/vignettes/design_meta_timestep.Rmd [hunk 7]
```diff
@@ -0,0 +1,403 @@
...
+$$
+The probability to arrive in any destination compartment is
+$$
+  P[E_{j,t+1}|S_t] = \frac{\lambda_j}{\sum_i\lambda_i}\left(1-e^{\sum_i\lambda_i}\right)
```
Comment (slwu89) [inline macro/vignettes/design_meta_timestep.Rmd]: Same thing, need a negative in the exponential.

**Review by slwu89 — COMMENTED** [review]
This is seriously cool. I only added comments/suggestions to the vignette, as I can't meaningfully critique the metaprogramming techniques.

File: macro/vignettes/design_meta_timestep.Rmd (hunk 4 above)
Comment (adolgert) [reply macro/vignettes/design_meta_timestep.Rmd]: I changed it to describe how many numerical operations it requires. I was thinking that there is a reason we don't always exponentiate. That can be slow.

**Pull request merged**